design_docs/generalized_horizontal_interpolation.md

```diff
@@ -0,0 +1,95 @@
+
+<h2>Requirement: First-order interpolation
+Date last modified: 2017/01/19
+Contributors: Xylar Asay-Davis</h2>
+
+The option to interpolate smoothly (e.g. linearly or with barycentric coordinates) between cell-centered values should be added.  The calling code should easily be able to select either nearest-neighbor or first-order interpolation with a flag as part of the initialization step is required and from then on should not need to keep track of which method is being used.
+
+<h2>Requirement: Interpolation should handle periodic boundaries
+Date last modified: 2017/01/19
+Contributors: Xylar Asay-Davis</h2>
+
+Interpolation code should be aware of periodic boundaries and should perform interpolation that is consistent with those boundaries.
+
+<h2>Consideration: Support caching results from any costly, one-time geometric computations
+Date last modified: 2017/01/20
+Contributors: Xylar Asay-Davis</h2>
+
+For many potential algorithms used to perform interpolation, there is likely to be a relatively costly step of computing fields such as indices into input data fields and interpolation weights that 1) only need to be computed once for a given input mesh and set of output points and 2) are independent of the data in the field being interpolated.  If this data were cached, it could mean that rerunning the analysis (which might be very desirable, e.g., while monitoring the progress of a run) would be much cheaper than the initial run.
+
+<h1> Algorithmic Formulations </h1>
+<h2>Design solution: Support for arbitrary output interpolation points
+Date last modified: 2017/01/19
+Contributors: Xylar Asay-Davis</h2>
+
+The calling code will be required to supply the interpolation points in cartesian space instead of supplying lon-lat bounds and steps.  This will allow greater generality.  A helper function will be provided that can produce a regularly spaced lon-lat grid.  Another helper function (lon_lat_to_cartesian) is already available for converting from lon-lat to Cartesian space.  Note: code calling using this helper function should use the spherical radius from an MPAS restart or input file instead of using the default value.
+The KD tree method currently used for nearest-neighbor interpolation does not care about how the interpolation points are spaced and thus will remain largely unchanged.
+
+<h2>Design solution: Cartesian input mesh
+Date last modified: 2017/01/19
+Contributors: Xylar Asay-Davis</h2>
+
+Currently, lonCell and latCell are passed as input arguments.  For efficiency, it would make more sense to use xCell, yCell, zCell rather than lonCell and latCell.  (Currently, fields equivalent to xCell, yCell and zCell are being computed from lonCell and latCell in the nearest neighbor code.)  Moreover, to support more generality in the interpolation algorithm (see next design solution), it may be best to pass an xarray data set containing xCell, yCell and zCell but also other mesh variables needed for linear interpolation.
```

Comment:
xCell should be noted as `xCell` in general to indicate model-specific variable.  This is true for other variables throughout.